sys/include/net/sock/dns.h

```diff
@@ -34,18 +34,37 @@ extern "C" {
 #endif
 
 /**
- * @brief DNS internal structure
+ * @brief DNS internal structure without payload
+ *
+ * @see   sock_dns_hdr_impl_t
  */
 typedef struct {
     uint16_t id;        /**< read           */
     uint16_t flags;     /**< DNS            */
     uint16_t qdcount;   /**< RFC            */
     uint16_t ancount;   /**< for            */
     uint16_t nscount;   /**< detailed       */
-    uint16_t arcount;   /**< explanations   */
-    uint8_t payload[];  /**< !!             */
+    uint16_t arcount;   /**< explanations!! */
 } sock_dns_hdr_t;
 
+#ifndef __cplusplus
+/**
+ * @brief   DNS internal structure with payload
+ * @extends sock_dns_hdr_t
+ * @details Auxiliary struct that contains a zero-length array as convenience
+ *          pointer to the payload. Only for use in C, invalid in ISO-C++.
+ */
+typedef struct {
+    uint16_t id;        /**< read           */
+    uint16_t flags;     /**< DNS            */
+    uint16_t qdcount;   /**< RFC            */
+    uint16_t ancount;   /**< for            */
+    uint16_t nscount;   /**< detailed       */
+    uint16_t arcount;   /**< explanations   */
+    uint8_t  payload[]; /**< !!             */
```

Comment:
could just this field be guarded by "__cplusplus"?

Reply:
Lol, I wasn't even thinking about this in #11299 :sweat_smile:

Reply:
Lol, yes of course.
Although when I think about it I like the expressiveness of a having separate struct